Add support for font-family: math.

eerii opened the pull request:
#### f5332f0677c1f712f9b63abdf717bed79a4ad854
<pre>
Add support for font-family: math.
<a href="https://bugs.webkit.org/show_bug.cgi?id=209595">https://bugs.webkit.org/show_bug.cgi?id=209595</a>

Reviewed by Frédéric Wang.

Add a new generic `math` font family following
<a href="https://www.w3.org/TR/css-fonts-4/#valdef-font-family-math.">https://www.w3.org/TR/css-fonts-4/#valdef-font-family-math.</a> Use the math
font list previously hardcoded in `mathml.css` to find a suitable option.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/generic-family-keywords-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/generic-family-keywords-003-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/default-font-family-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/default-font-family-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/default-font-family-expected.txt:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::familyNameFromPrimitive):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::mathFontList):
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/WebKitFontFamilyNames.in:
* Source/WebCore/css/mathml.css:
(math):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
(WebCore::CSSPropertyParserHelpers::genericFontFamily):
(WebCore::CSSPropertyParserHelpers::genericFontFamilyIndex):
(WebCore::CSSPropertyParserHelpers::consumeGenericFamilyUnresolved):
* Source/WebCore/platform/graphics/FontGenericFamilies.cpp:
(WebCore::FontGenericFamilies::fontFamily const):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/style/StyleExtractorConverter.h:
(WebCore::Style::ExtractorConverter::convertFontFamily):
* Source/WebCore/style/StyleExtractorSerializer.h:
(WebCore::Style::ExtractorSerializer::serializeFontFamily):
* Source/WebInspectorUI/UserInterface/Models/CSSKeywordCompletions.js:

Canonical link: <a href="https://commits.webkit.org/299006@main">https://commits.webkit.org/299006@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1f122f4a1b83b67c92b78f219f3df4ff733a8cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116528 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122587 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67092 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118417 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36883 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44777 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88485 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/42947 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119477 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104532 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68930 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28477 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22641 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66266 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98788 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22799 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125734 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43422 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97188 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100741 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96983 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24844 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42287 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20200 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39376 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43310 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48901 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Checked out pull request; Found 70326 issues") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42776 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46116 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44481 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ca2b46e45593167458bebff6db756820a57cf981. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48510

fred-wang commented on LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-fonts/generic-family-keywords-002-expected.txt:
```diff
@@ -7,7 +7,7 @@ FAIL font-family: -webkit-cursive treated as <font-family>, not <generic-name> a
 FAIL font-family: -webkit-monospace treated as <font-family>, not <generic-name> assert_equals: expected 50 but got 30
 FAIL font-family: -webkit-system-ui treated as <font-family>, not <generic-name> assert_equals: expected 50 but got 30
-PASS font-family: -webkit-math treated as <font-family>, not <generic-name>
+FAIL font-family: -webkit-math treated as <font-family>, not <generic-name> assert_equals: expected 50 but got 25
```
It's a pity we expose -webkit-math...

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -2435,6 +2457,43 @@ void webkit_settings_set_pictograph_font_family(WebKitSettings* settings, const
+ */
+const gchar* webkit_settings_get_math_font_family(WebKitSettings* settings)
+{
+    g_return_val_if_fail(WEBKIT_IS_SETTINGS(settings), 0);
```
I think some of the getters in this file still use `0` because some of this code is quite old, from the times before `nullptr` was a thing. We prefer the latter nowadays:

```suggestion
    g_return_val_if_fail(WEBKIT_IS_SETTINGS(settings), nullptr);
```

aperezdc commented on Source/WebCore/testing/InternalSettings.cpp:
```diff
@@ -280,6 +284,18 @@ ExceptionOr<void> InternalSettings::setPictographFontFamily(const String& family
+    if (!m_page)
+        return Exception { ExceptionCode::InvalidAccessError };
+    UScriptCode code = scriptNameToCode(script);
+    if (code == USCRIPT_INVALID_CODE)
+        return { };
```
Maybe I am missing something here; but instead of silently doing nothing, wouldn't it be better to do an `Exception`-return when the passed argument is an invalid script name?

aperezdc replied:
Right, the other font families indeed do the same. On second thought, I suppose it doesn't matter much that the case is a no-op, because then the font is not really changed and that would make tests fail because the expectations would not be matched 🤔

aperezdc reviewed (commented):
I left a couple of small comments.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c6a9c18710d8218c99e1587975570115d061e242. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48510

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c6a9c18710d8218c99e1587975570115d061e242. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48510

fred-wang commented:
> This PR is now split into two

Hey, not sure whether you did that or not, but maybe you can consider keeping a minimal amount of reviewers for this change (I see 8 of them).

eerii commented:
> Hey, not sure whether you did that or not, but maybe you can consider keeping a minimal amount of reviewers for this change (I see 8 of them).

I'm not sure exactly why that happened, the only reviewer I manually added was you. It may be because of the codeowners feature, and since this change has a lot of boilerplate in different files it may be triggering that.

> I relaunched the CI so let's see.

It's still failing quite a bit. I'm going to do a clean build and do some tests locally, because there may still be something wrong with font selection.

eerii commented:
I opened #49208 which adds a new preference for configuring the generic font family.

As for this PR, almost all of the previous test failures are now resolved. The only one missing seems to be `TestWebKitAPI.WebKit.SaveDataToFile` in `api-mac`, but I think it might not be related. I'd rerun it but I don't have privileges. I'm not quite sure why `win-tests` was cancelled either.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e1f122f4a1b83b67c92b78f219f3df4ff733a8cf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48510

eerii committed:
Add support for font-family: math.
https://bugs.webkit.org/show_bug.cgi?id=209595

Reviewed by Frédéric Wang.

Add a new generic `math` font family following
https://www.w3.org/TR/css-fonts-4/#valdef-font-family-math. Use the math
font list previously hardcoded in `mathml.css` to find a suitable option.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/generic-family-keywords-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/generic-family-keywords-003-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/mathml/relations/css-styling/default-font-family-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/mathml/relations/css-styling/default-font-family-expected.txt:
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/mathml/relations/css-styling/default-font-family-expected.txt:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::CSSFontFaceSet::familyNameFromPrimitive):
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::mathFontList):
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/WebKitFontFamilyNames.in:
* Source/WebCore/css/mathml.css:
(math):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
(WebCore::CSSPropertyParserHelpers::genericFontFamily):
(WebCore::CSSPropertyParserHelpers::genericFontFamilyIndex):
(WebCore::CSSPropertyParserHelpers::consumeGenericFamilyUnresolved):
* Source/WebCore/platform/graphics/FontGenericFamilies.cpp:
(WebCore::FontGenericFamilies::fontFamily const):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/style/StyleExtractorConverter.h:
(WebCore::Style::ExtractorConverter::convertFontFamily):
* Source/WebCore/style/StyleExtractorSerializer.h:
(WebCore::Style::ExtractorSerializer::serializeFontFamily):
* Source/WebInspectorUI/UserInterface/Models/CSSKeywordCompletions.js:

Canonical link: https://commits.webkit.org/299006@main
